Shell/AST.h

```diff
@@ -0,0 +1,730 @@
+    virtual ~GlobValue();
+    virtual bool is_glob() const override { return true; }
+    GlobValue(String glob)
+        : m_glob(glob)
+    {
+    }
+
+private:
+    String m_glob;
+};
+
+class SimpleVariableValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override;
+    virtual ~SimpleVariableValue();
+    // FIXME: Should override is_list and is_string.
+    SimpleVariableValue(String name)
+        : m_name(name)
+    {
+    }
+
+private:
+    String m_name;
+};
+
+class SpecialVariableValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override;
+    virtual ~SpecialVariableValue();
+    // FIXME: Should override is_list and is_string.
+    SpecialVariableValue(char name)
+        : m_name(name)
+    {
+    }
+
+private:
+    char m_name;
+};
+
+class TildeValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override;
+    virtual ~TildeValue();
+    virtual bool is_string() const override { return true; }
+    TildeValue(String name)
+        : m_username(name)
+    {
+    }
+
+private:
+    String m_username;
+};
+
+class Node : public RefCounted<Node> {
+public:
+    virtual void dump(int level) = 0;
+    virtual RefPtr<Value> run(TheExecutionInputType) = 0;
+    virtual void highlight_in_editor(Line::Editor&, RefPtr<Shell>, HighlightMetadata = {}) = 0;
+    virtual Vector<Line::CompletionSuggestion> complete_for_editor(RefPtr<Shell>, size_t, RefPtr<Node> matching_node);
+    Vector<Line::CompletionSuggestion> complete_for_editor(RefPtr<Shell> shell, size_t offset);
+    virtual HitTestResult hit_test_position(size_t offset)
+    {
+        if (m_position.in_range(offset))
+            return { this, nullptr };
+        return { nullptr, nullptr };
+    }
+    virtual String class_name() const { return "Node"; }
+    Node(Position);
+    virtual ~Node();
+
+    virtual bool is_bareword() const { return false; }
+    virtual bool is_command() const { return false; }
+    virtual bool is_execute() const { return false; }
+    virtual bool is_glob() const { return false; }
+    virtual bool is_variable_decl() const { return false; }
+
+    virtual bool is_list() const { return false; }
+
+    const Position& position() const { return m_position; }
+
+protected:
+    Position m_position;
+};
+class And : public Node {
+public:
+    virtual void dump(int level) override;
+    virtual RefPtr<Value> run(TheExecutionInputType) override;
+    virtual void highlight_in_editor(Line::Editor&, RefPtr<Shell>, HighlightMetadata = {}) override;
+    virtual HitTestResult hit_test_position(size_t) override;
+    virtual String class_name() const override { return "And"; }
+    And(Position, RefPtr<Node>, RefPtr<Node>);
+    virtual ~And();
+
+private:
+    RefPtr<Node> m_left;
+    RefPtr<Node> m_right;
+};
+
+class PrependToList : public Node {
```

Comment:
Is this the space operator? If so, it doesn't just prepend, it concats — you can use it to append (an element to a list), or to indeed concat two lists.

Reply:
prepend a list to another? what's wrong with that name?
it does do what you described, it actually flattens two things and prepends one result to the other one

Reply:
`#incude <english-is-not-my-first-language.h>`

To my ear, "prepend to list" (without clarifying that we're prepending *another list*) sounds like we're prepending *one element* to a list.

Reply:
the parser calls it as an element too, but it doesn't really matter here, everything is flattened.

aside, does `append()` sound wrong for lists too? in which case, give whomever came up with `Vector::append` a piece of your mind :)

Reply:
> aside, does `append()` sound wrong for lists too? in which case, give whomever came up with `Vector::append` a piece of your mind :)

*append a list to another list* (so, `Vector<T>::append(Vector<T>&&)`) sounds fine, I guess, if I see the type. If I don't see the type, I'd prefer it the way it's done in Python — `append()` for an element, `extend()` for adding items from another list.